api/proto/info/info.proto

```diff
@@ -393,6 +394,37 @@ message ZCellularProvider {
   bool roaming = 4;
 }
 
+// Snapshot states
+enum ZSnapshotState {
+  Z_SNAPSHOT_STATE_UNSPECIFIED = 0;
+  // This state is used when a snapshot is in the process of being
+  // created or an error occurred during the first attempt to create it.
+  // (For example, the operation was delayed)
+  Z_SNAPSHOT_STATE_CREATING = 1;
+  // This state is used when the snapshot has been successfully created.
+  Z_SNAPSHOT_STATE_CREATED = 2;
+  // This state is used when the snapshot is pending deletion or
+  // the first deletion attempt was not successful.
+  Z_SNAPSHOT_STATE_DELETING = 3;
+  // This state is used to send information to the controller about a
+  // snapshot that was implicitly deleted after a rollback snapshot
+  // or volume delete command.
```

Comment:
If the volume is deleted it means all snapshots for that volume are also deleted. If we are going to include those as IMPLICTLY_DELETED, then for how long time do you expect EVE-OS to keep sending them as IMPLICITLY_DELETED?
It might be simpler to just omit snapshots for volumes which have been deleted from the info message.

Reply:
> If the volume is deleted it means all snapshots for that volume are also deleted. If we are going to include those as IMPLICTLY_DELETED, then for how long time do you expect EVE-OS to keep sending them as IMPLICITLY_DELETED?

I'm assuming that after the rollback we report that the snapshot has been implicitly deleted, and the next time we don't expect to get a SnapshotConfig for it.

> It might be simpler to just omit snapshots for volumes which have been deleted from the info message.

I think we can't rely on the controller to know that newer snapshots have been deleted on rollback.
If the controller deletes older snapshots before the rollback, then reporting implicit deletion will not break the logic, but if the controller does not delete it, then it will be critical.

Reply:
> I'm assuming that after the rollback we report that the snapshot has been implicitly deleted, and the next time we don't expect to get a SnapshotConfig for it.

It doesn't make sense for EVE-OS to make such an assumption. For one reason there can be races where you report the info with the implicitly deleted at the same time as you fetch the EdgeDevConfig. And with network outages there could be an hour or a week from processing a rollback until the info and a subsequent config are handled. Last but not least, EVE-OS should be robust even if the controller doesn't remove the snapshots from the config even after a rollback. (I'd expect a well-behaved controller to remove them as part of putting the rollbackcmd in the config, but EVE-OS code should not assume such things.)

Reply:
If we want to avoid a race condition for such a case, we can add a condition an additional check for the presence of a logical volume, before sending "implicitly deleted", and if the logical volume for the snapshot does not already exist, then do not send a message with such a status.

As for the rest, we can't delete what is already implicitly deleted, so EVE will mark messages as deleted and not notify about them in the next informational messages, but if we still continue to receive configurations for implicitly deleted messages, then it is likely on the controller that something went wrong and we should probably send an update on snapshots that are no longer there.

In any case, this should not affect the stability of EVE, except that there is a question of informing.

Wouldn't this be a reliable option for EVE-OS?

Reply:
This isn't part of the API definition but the EVE-OS implementation but I'm asking to make sure we won't hit issues in the implementation. If a controller is suboptimal and does not remove the newer SnapshotConfig when doing a rollback to something old.
For example, start with uuid=1 for volume_uuid=X, then add a second SnapshotConfig with uuid=2 for volume_uuid=X and then later it can send a RollbackCmd with snapshot_uuid=1, and the SnapshotConfig with uuid=2 for volume_uuid=X remains. How does EVE-OS know that it has already done (and implicitly deleted) uuid=2? Does it need to persist the implicitly deleted SnapshotConfig across reboots so that it can tell it should ignore that stale uuid=2 from the controller?